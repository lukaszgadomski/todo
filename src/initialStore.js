export const filterTodoType = {
  ALL: 0,
  ACTIVE: 1,
  COMPLETED: 2
};

export default {
  //group component state
  groups: {
    search: "",
    list: [],
    todos_by_group_id: {}, //list of todos for each list by group id
    selected_group_id: null
  },
  //todos component state
  todos: {
    search: "",
    list: null, //todos from selected group, if it equals null then no group is selected
    filter: filterTodoType.ALL
  }
};
