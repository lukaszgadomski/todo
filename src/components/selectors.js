/***
 * get selected group
 * filter todos by filter type
 * filter todos by search text if it's not empty
 */
import { filterTodoType } from "../initialStore";
import { createSelector } from "reselect";

export const getGroups = state => state.groups.list;
export const getGroupsSearch = state => state.groups.search;
export const getTodosByGroupId = state => state.groups.todos_by_group_id;
export const getSelectedGroupId = state => state.groups.selected_group_id;
export const getTodosFilter = state => state.todos.filter;
export const getTodosSearch = state => state.todos.search;

export const searchPredicate = search => field => item => {
  if (!search || !search.length) {
    return true;
  }
  return item[field].indexOf(search) !== -1;
};

// export const filterTypePredicate = filter => ;

export const getSelectedTodosSelector = createSelector(
  getSelectedGroupId,
  getTodosByGroupId,
  (groupId, groupsById) => (groupId ? groupsById[groupId] : null)
);

export const getSelectedGroup = createSelector(
  getGroups,
  getSelectedGroupId,
  (groups, groupId) => {
    if (!groupId) {
      return null;
    }
    return groups.find(group => group.id === groupId);
  }
);

//returns todos list: selected, filtered by search and status
export const getVisibleTodos = createSelector(
  getSelectedTodosSelector,
  getTodosFilter,
  getTodosSearch,
  (todos, filterType, searchText) => {
    if (todos == null) {
      return [];
    }
    //simply return todos if filters are empty
    if (filterType === filterTodoType.All && searchText === "") {
      return todos;
    }
    const search = searchPredicate(searchText)("name");
    return todos.filter(todo => {
      return (
        search(todo) &&
        (filterType === filterTodoType.ALL || filterType === todo.state)
      );
    });
  }
);

//return groups with todos count
export const getVisibleGroups = createSelector(
  getGroups,
  getSelectedTodosSelector,
  getGroupsSearch,
  (groups, todos, searchText) => {
    if (searchText === "") {
      return groups;
    }
    return groups.filter(searchPredicate(searchText)("name"));
  }
);
