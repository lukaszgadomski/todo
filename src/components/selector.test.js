import { getVisibleTodos, getVisibleGroups } from "./selectors";
import { filterTodoType } from "../initialStore";

const stringy = obj => JSON.stringify(obj, null, 4);

let state = {};
beforeEach(() => {
  state = {
    groups: {
      search: "",
      list: [{ name: "List n. 1", count: 4 }, { name: "List n. 2", count: 1 }],
      todos_by_group_id: {
        1: [
          {
            id: 1,
            group_id: 1,
            name: "bob clean",
            status: filterTodoType.COMPLETED
          },
          {
            id: 1,
            group_id: 1,
            name: "bob the",
            status: filterTodoType.ACTIVE
          },
          { id: 1, group_id: 1, name: "code", status: filterTodoType.ACTIVE },
          { id: 1, group_id: 1, name: "do it", status: filterTodoType.ACTIVE }
        ],
        2: [{ name: "i'm in second list!", status: filterTodoType.COMPLETED }]
      }, //list of todos for each list by group id
      selected_group_id: 1
    },
    //todos component state
    todos: {
      search: "bob",
      //    list: null, //todos from selected group, if it equals null then no group is selected
      filter: filterTodoType.ACTIVE
    }
  };
});

describe("getVisibleTodos", () => {
  test("filter by status and text and group id", () => {
    expect(getVisibleTodos(state)).toMatchSnapshot();
  });
});

describe("getVisibleGroups", () => {
  test("filter by text ", () => {
    expect(getVisibleGroups(state)).toMatchSnapshot();
  });
});
