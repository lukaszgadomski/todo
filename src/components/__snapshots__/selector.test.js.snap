// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getVisibleGroups filter by text  1`] = `
Array [
  Object {
    "count": 4,
    "name": "List n. 1",
  },
  Object {
    "count": 1,
    "name": "List n. 2",
  },
]
`;

exports[`getVisibleTodos filter by status and text and group id 1`] = `
Array [
  Object {
    "group_id": 1,
    "id": 1,
    "name": "bob the",
    "status": 1,
  },
]
`;
